### PR TITLE
SDI-287 refactor prisoner difference service

### PR DESCRIPTION
**Pull request by mikehalmamoj**
Previously I experimented with using standalone functions instead of a full Spring service. But as soon dependencies are needed everything gets awkward, so I've introduced a proper Spring singleton for PrisonerDifferenceService.
Also moved all prisoner difference related functionality into the new service as too much detail was exposed in the PrisonerIndexService.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/PrisonerIndexService.kt [hunk 1]
```diff
@@ -101,54 +97,11 @@ class PrisonerIndexService(
       }
     }
 
-    generateDiffTelemetry(existingPrisoner, offenderBooking, storedPrisoner)
-    generateDiffEvent(existingPrisoner, offenderBooking, storedPrisoner)
+    prisonerDifferenceService.handleDifferences(existingPrisoner, offenderBooking, storedPrisoner)
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/PrisonerIndexService.kt:100]: Moved all prisoner diff related code into the service.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/PrisonerIndexServiceTest.kt [hunk 2]
```diff
@@ -230,226 +223,4 @@ class PrisonerIndexServiceTest {
       latestLocationId = "MDI"
     )
   }
-
-  @Nested
-  inner class SyncPrisonerDifferencesTelemetry {
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/PrisonerIndexServiceTest.kt:235]: Tests also moved to service to follow code.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/diff/PrisonerEventHashRepositoryTest.kt [hunk 3]
```diff
@@ -96,6 +97,6 @@ class PrisonerEventHashRepositoryTest : IntegrationTest() {
 
     saved = find("A2222AA")
     assertThat(saved?.prisonerHash).isEqualTo(223)
-    assertThat(saved?.updatedDateTime).isEqualTo(updateTime)
+    assertThat(saved?.updatedDateTime).isEqualTo(updateTime2)
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/diff/PrisonerEventHashRepositoryTest.kt:100]: Improve test from previous PR.

**Review by lbennett-moj — APPROVED** [review]
LGTM

**Pull request merged**